resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/SparkRackResolver.scala

```diff
@@ -17,24 +17,90 @@
 
 package org.apache.spark.deploy.yarn
 
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+import com.google.common.base.Strings
 import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.CommonConfigurationKeysPublic
+import org.apache.hadoop.net._
+import org.apache.hadoop.util.ReflectionUtils
 import org.apache.hadoop.yarn.util.RackResolver
 import org.apache.log4j.{Level, Logger}
 
+import org.apache.spark.internal.Logging
+
 /**
- * Wrapper around YARN's [[RackResolver]]. This allows Spark tests to easily override the
+ * Re-implement YARN's [[RackResolver]]. This allows Spark tests to easily override the
  * default behavior, since YARN's class self-initializes the first time it's called, and
  * future calls all use the initial configuration.
  */
-private[yarn] class SparkRackResolver {
+private[spark] class SparkRackResolver {
 
   // RackResolver logs an INFO message whenever it resolves a rack, which is way too often.
   if (Logger.getLogger(classOf[RackResolver]).getLevel == null) {
     Logger.getLogger(classOf[RackResolver]).setLevel(Level.WARN)
   }
 
   def resolve(conf: Configuration, hostName: String): String = {
-    RackResolver.resolve(conf, hostName).getNetworkLocation()
+    SparkRackResolver.coreResolve(conf, List(hostName)).head.getNetworkLocation
+  }
+
+  /**
+   * Added in SPARK-27038.
+   * This should be changed to `RackResolver.resolve(conf, hostNames)`
+   * in hadoop releases with YARN-9332.
+   */
+  def resolve(conf: Configuration, hostNames: List[String]): List[Node] = {
+    SparkRackResolver.coreResolve(conf, hostNames)
   }
+}
 
+/**
+ * Utility to resolve the rack for hosts in an efficient manner.
+ * It will cache the rack for individual hosts to avoid
+ * repeatedly performing the same expensive lookup.
+ *
+ * Its logic refers [[org.apache.hadoop.yarn.util.RackResolver]] and enhanced.
+ * This will be unnecessary in hadoop releases with YARN-9332.
+ * With that, we could just directly use [[org.apache.hadoop.yarn.util.RackResolver]].
+ * In the meantime, this is a re-implementation for spark's use.
+ */
+object SparkRackResolver extends Logging {
+  private var dnsToSwitchMapping: DNSToSwitchMapping = _
+  private var initCalled = false
+
+  def coreResolve(conf: Configuration, hostNames: List[String]): List[Node] = {
+    if (!initCalled) {
```

Comment:
This doesn't look thread-safe. You can have this called from at least two different threads in cluster mode (the task scheduler and the YarnAllocator).

I just want to say again that I dislike this object, because caching things in static fields is generally a bad idea. In this case it allows you to share that state between the two callers I mention above, but it also means that if you start multiple Spark apps in the same JVM, the first one will be the one that defined what this configuration will be.

I'll let this object issue slide for now since that is less common of a use case (and the underlying YARN code already had this problem before this change anyway), but the thread-safety needs to be fixed.

Reply:
Why I use static fields here is for the cache `dnsToSwitchMapping`. I hope only one `dnsToSwitchMapping` in JVM since there is a hashmap `cache` in this instance. I don't want to keep more than one cache hashmap in on JVM.

Reply:
I've added the sync block to fix thread-safe problem.

Reply:
As Bruce mentioned this is not thread-safe. You want:

```
  @volatile private var dnsToSwitchMapping: DNSToSwitchMapping = _

  private def ensureInitialized(): Unit = {
    if (dnsToSwitchMapping == null) {
      synchronized {
        if (dnsToSwitchMapping == null) {
          initialize(conf)
        }
      }
    }
  }
```

You don't need the boolean since you always initialize `dnsToSwitchMapping`.

> Why I use static fields here is for the cache dnsToSwitchMapping

I know. What I'm saying is that it's generally a bad idea. e.g. you can't test this with different configurations, for example, because once the first test runs, this is initialized forever.

I'm not saying caching is bad. I'm saying using static variables for caching is bad. But anyway, this is already a problem because YARN does that sort of thing, so this is no worse than before.

But since we're here, you could:

- move all this code to the SparkResolver class, do initialization in the constructor
- have the object cache a `SparkResolver` object

```
object SparkResolver {

  private var instance: SparkResolver = _

  def get(conf: Configuration): SparkResolver = synchronized { ... }

}
```

That's less icky because it at least allows you to create a new resolver with a different config if you really want to.

Reply:
Every calling of `get` in different threads will generate a new SparkResolver since the `conf` instances are different. So how to keep only one cache in memory?

Reply:
Meanwhile, since `var instance: SparkResolver` is static,  how to maintain more than one SparkResolvers for different `conf`?

Reply:
Let me fix the thread-safe problem first.

Reply:
The `get` in an object is a compromise, if you didn't understand.

It allows the caching you want while keeping the logic where it belongs, in the class. Everybody who uses the object will get the cached version; the cached instance is initialized in the first call. Just like your code here, but much simpler.

And it allows those who want to instantiate a separate resolver with a separate config if they so choose.

Reply:
I think I understand now. I will update it

Reply:
Refactored.